staging/src/k8s.io/api/bulk/v1alpha1/types.go

```diff
@@ -0,0 +1,206 @@
+	// Null for messages issued by server (WatchEvent, WatchStopped triggered by server)
+	// +optional
+	RequestID *string `json:"requestId,omitempty" protobuf:"bytes,1,opt,name=requestId"`
+
+	// Contains error details when request was failed.
+	// +optional
+	Failure *metav1.Status `json:"failure,omitempty" protobuf:"bytes,2,opt,name=failure"`
+
+	// Sent when watch was manually stopped.
+	// +optional
+	WatchStopped *WatchStopped `json:"watchStopped,omitempty" protobuf:"bytes,3,opt,name=watchStopped"`
+
+	// Sent when new watch was started.
+	// +optional
+	WatchStarted *WatchStarted `json:"watchStarted,omitempty" protobuf:"bytes,4,opt,name=watchStarted"`
+
+	// Contains single watch event.
+	// +optional
+	WatchEvent *BulkWatchEvent `json:"watchEvent,omitempty" protobuf:"bytes,5,opt,name=watchEvent"`
+
+	// Contains single object.
+	// +optional
+	GetResult *GetResult `json:"getResult,omitempty" protobuf:"bytes,6,opt,name=getResult"`
+
+	// Contains single object.
+	// +optional
+	ListResult *ListResult `json:"listResult,omitempty" protobuf:"bytes,7,opt,name=listResult"`
+}
+
+// GroupVersionResource identifies resource.
+type GroupVersionResource struct {
+	// Name of APIGroup, defaults to core API group.
+	// +optional
+	Group string `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+
+	// Version of APIGroup, defaults to preferred version.
+	// +optional
+	Version string `json:"version,omitempty" protobuf:"bytes,2,opt,name=version"`
+
+	// Resource identifier (pods, events etc).  Note: this is not the kind.
+	Resource string `json:"resource" protobuf:"bytes,3,opt,name=resource"`
+}
+
+// ItemSelector is the query options to get/watch calls.
+type ItemSelector struct {
+	// Resource identifies requested resource.
+	GroupVersionResource `json:",inline" protobuf:"bytes,1,name=resource"`
+
+	// Name is the object name.
+	Name string `json:"name,omitempty" protobuf:"bytes,5,opt,name=namespace"`
+
+	// Namespace
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,4,opt,name=namespace"`
+
+	// Query options used to filter watched resources.
+	// Field 'Watch' is ignored by bulk api.
+	// +optional
+	Options *metav1.GetOptions `json:"options" protobuf:"bytes,6,name=options"`
+}
+
+// ListSelector is the query options to list/watchList calls.
+type ListSelector struct {
+	// Resource identifies requested resource.
+	GroupVersionResource `json:",inline" protobuf:"bytes,1,name=resource"`
+
+	// Namespace
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,4,opt,name=namespace"`
+
+	// Query options used to filter watched resources.
+	// Field 'Watch' is ignored by bulk api.
+	// +optional
+	Options *metav1.ListOptions `json:"options" protobuf:"bytes,2,name=options"`
+}
+
+// BulkWatchEvent represents a single event to a watched resource.
+type BulkWatchEvent struct {
+	// Identifier of the watch.
+	WatchID string `json:"watchId" protobuf:"bytes,1,name=watchId"`
```

Comment:
Do we need both WatchID and RequestID?

Reply:
Yes, we need both.
WatchID might be used by client to map "watch options" & watch events.
RequestID is used to map async requests & responses (it is possible to send a new request before server answered to the previous one).